jupyter_notebook/analytics.py

```diff
@@ -0,0 +1,194 @@
+        self.data = {}
+        self.ioloop = IOLoop.current()
+
+    def control_info(self):
+        if self._control_info is None:
+            try:
+                with open(self.control_file, 'r') as f:
+                    self._control_info = json.load(f)
+            except OSError as e:
+                if e.errno == errno.ENOENT:
+                    self._control_info = {}
+                else:
+                    raise
+
+        return self._control_info
+
+    def store_control_info(self):
+        assert self._control_info is not None
+        try:
+            os.makedirs(self.directory)
+        except OSError as e:
+            if e.errno != errno.EEXIST:
+                raise
+
+        with open(self.control_file, 'w') as f:
+            json.dump(self._control_info, f)
+
+    def read_data(self):
+        try:
+            with open(self.data_file, 'r') as f:
+                self.data = json.load(f)
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise
+
+    def store_data(self):
+        try:
+            os.makedirs(self.directory)
+        except OSError as e:
+            if e.errno != errno.EEXIST:
+                raise
+
+        with open(self.data_file, 'w') as f:
+            json.dump(self.data, f)
+
+    def consent_status(self):
+        ci = self.control_info()
+        if 'consent' in ci:
+            return CONSENT_OK if ci['consent'] else CONSENT_DENIED
+        return CONSENT_UNKNOWN
+
+    def change_consent_status(self, status):
+        ci = self.control_info()
+        if status == CONSENT_OK:
+            if not ci.get('consent', False):
+                self.enable()
+            ci['consent'] = True
+        elif status == CONSENT_DENIED:
+            ci['consent'] = False
+        else:
+            ci.pop('consent', None)
+
+        self.store_control_info()
+
+    def last_submission(self):
+        ci = self.control_info()
+        if 'last_submission' in ci:
+            return parse_date(ci['last_submission'])
+
+        return None
+
+    def startup(self):
+        webapp = self.nbapp.web_app
+
+        consent = self.consent_status()
+        if consent == CONSENT_OK:
+            self.enable()
+        elif consent == CONSENT_UNKNOWN:
+            webapp.settings['analytics_prompt'] = True
+
+        base_url = webapp.settings['base_url']
+        webapp.add_handlers(".*$", [
+            (ujoin(base_url, r"/analytics/consent"), ConsentHandler,
+                {'sender': self}),
+        ])
+
+    def enable(self):
+        self.enabled = True
+        self.nbapp.log.info("Enabling analytics")
+        self.read_data()
+        self.record_version('jupyter_notebook', __version__)
+        self.record_version('server_python', sys.version)
+        self.schedule_send()
+
+        self.socket = zmq.Context.instance().socket(zmq.PULL)
+        port = self.socket.bind_to_random_port('tcp://127.0.0.1')
```

Comment:
I was assuming so, but it wasn't immediately obvious where to get the config values from. `NotebookApp.ip` is the IP address for the HTTP server, I think, not for the ZMQ sockets. I guess it's configured on KernelManager, but fishing it out of the config seems ugly.